### ISSUE TITLE
Avoid IR traversal by converting (some of) our `Pass`es to mini passes

### ISSUE THREAD

**Issue by JaroslavTulach** [enso-org/enso#10981]
After discussing this idea @hubertp pointed out there is a [prior work](https://dl.acm.org/doi/10.1145/3140587.3062346) - read at [Miniphases: compilation using modular and efficient tree transformations](https://infoscience.epfl.ch/server/api/core/bitstreams/8ab72c0a-8aa6-4dee-a704-3504938dc316/content).

### Current State

Enso static compiler currently works on `IR` tree. The work is performed by many [IR passes](https://github.com/enso-org/enso/blob/77183e50e941dbeb00b384ed56a88be68fb30088/engine/runtime-compiler/src/main/scala/org/enso/compiler/Passes.scala#L29). Each `Pass` gets `IR` tree as an input and is supposed to return copy of the `IR` tree as output. This is very inefficient:
- most of the passes aren't interested in 99% of the `IR` tree - yet they have to traverse it themselves
- a copy of `IR` tree _escapes_ (from a point of _escape analysis_) - e.g. it has to be allocated on the heap - as it is too large/heterogenous to optimize by any compiler
- by allocating copies of the whole `IR` on the heap we create unnecessary pressure on the GC 

### Goal

Rather than passing the whole `IR` as data and letting each `Pass` to implement the traversal itself, let's:
- rewrite eligible passes to simple `PartialFunction<IR, IR>`
- chain these passes into huge mega `PartialFunction<IR, IR>`
- create a general pass to traverse the IR (**leafs first**) and apply the _mega partial function_ on each `IR` element

By doing this we:
- eliminate unnecessary traversal - the whole IR tree is going to be scanned just once for the _mega partial function_
- give _escape analysis a chance_ to virtualize object allocation inside of the _mega partial function_ lowering the need to allocate objects on heap

### Impact

A lot of `Pass`es may be eligible for the _mini passes_ approach. Usually they just invoke `transformExpression` with a single partial function - all of them are eligible for rewrite:
- [AmbiguousImportsAnalysis](https://github.com/enso-org/enso/blob/77183e50e941dbeb00b384ed56a88be68fb30088/engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/AmbiguousImportsAnalysis.scala#L36)
- [BindingAnalysis](https://github.com/enso-org/enso/blob/77183e50e941dbeb00b384ed56a88be68fb30088/engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/BindingAnalysis.scala#L31)
- [CachePreferenceAnalysis](https://github.com/enso-org/enso/blob/77183e50e941dbeb00b384ed56a88be68fb30088/engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/CachePreferenceAnalysis.scala#L36)
- [DataflowAnalysis](https://github.com/enso-org/enso/blob/77183e50e941dbeb00b384ed56a88be68fb30088/engine/runtime-compiler/src/main/scala/org/enso/compiler/pass/analyse/DataflowAnalysis.scala#L52)
- `DemandAnalysis`
- `GatherDiagnostics`
- `ImportSymbolAnalysis`
- `TailCall`
- `OperatorToFunction`
- etc.

### Pre-requisite

We need _generic traversal function_. Right now there is `ir.Expression.transformExpression`, but we need such a `transform` function for every `IR` element. See:

- https://github.com/enso-org/enso/issues/10537#issuecomment-2230355981

**Update**: At the end #11191 uses `IR.mapExpressions` to traverse the tree and things seem to work fine.

**Comment by JaroslavTulach** [comment]
When at it please design the traversing architecture to be ready for parallelization by [fork join pool](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/util/concurrent/ForkJoinPool.html).

**Comment by JaroslavTulach** [comment]
Hello Pavel, I have a problem understanding following part of your standup:

>     * `Expression.withNewChildren` cannot be implemented via `mapExpression`.
>       * `Application.Prefix.mapExpressions` is recursive!

As far as I see, there is no recursion in `Application.Prefix.mapExpressions`. Here is my example:
```diff
enso.master$ git diff
diff --git engine/runtime-parser/src/test/java/org/enso/compiler/core/EnsoParserTest.java engine/runtime-parser/src/test/java/org/enso/compiler/core/EnsoParserTest.java
index 256d415952..2596841789 100644
--- engine/runtime-parser/src/test/java/org/enso/compiler/core/EnsoParserTest.java
+++ engine/runtime-parser/src/test/java/org/enso/compiler/core/EnsoParserTest.java
@@ -37,6 +37,27 @@ public class EnsoParserTest {
     """, true, true, true);
   }
 
+  @Test
+  public void testApplicationShallowTraversal() {
+    var ir = compile("""
+    fn x y z = x ((y 2) (z 1))
+    """);
+    java.util.function.BiConsumer<String, IR> log =
+        (prefix, fst) -> {
+          System.out.println(prefix + fst.getClass().getSimpleName() + ":" + fst.showCode());
+        };
+    ir.mapExpressions(
+        (fst) -> {
+          log.accept("first level: ", fst);
+          fst.mapExpressions(
+              (snd) -> {
+                log.accept("  second level: ", snd);
+                return snd;
+              });
+          return fst;
+        });
+  }
+
   @Test
   public void testLocationsApplicationsAndMethodCalls() {
     parseTest("""
```
it runs and prints:
```
first level: Prefix:((x) ((((y) 2)) ((z) 1)))
  second level: Literal:x
  second level: Prefix:((((y) 2)) ((z) 1))
```
e.g. **only first and second** level is iterated. Not the whole tree composed of `Application.Prefix` elements. If you really believe some implementation of `mapExpressions` is recursive - can you please support such claim with an example that demonstrates the recursiveness?

> `Expression.withNewChildren` cannot be implemented via `mapExpression`.

Is this claim based on the previous (not proved) one? Or is there some other reason why it doesn't seem possible?

**Comment by Akirathan** [comment]
> Is this claim based on the previous (not proved) one? Or is there some other reason why it doesn't seem possible?

@JaroslavTulach Sorry for not being precise. Let's look closely at your example.

Put
```scala
  "Deep traversal" should {
    "XX work" in {
      implicit val ctx: ModuleContext = mkModuleContext

      def log(prefix: String, fst: IR) = {
        println(prefix + fst.getClass.getSimpleName + ":" + fst.showCode())
      }

      val ir =
        """
          |fn x y z = x ((y 2) (z 1))
          |""".stripMargin.preprocessModule.analyse
      // prefix is `x ((y 2) (z 1))` expression
      val prefix = ir.bindings(0)
        .asInstanceOf[Method.Explicit]
        .body
        .asInstanceOf[Function.Lambda]
        .body
        .asInstanceOf[Application.Prefix]
      prefix.function.asInstanceOf[Name.Literal].name shouldEqual "x"
      prefix.arguments should have size 1
      val arg = prefix.arguments(0).asInstanceOf[CallArgument.Specified]
      arg.value.isInstanceOf[Application.Prefix] shouldBe true
      prefix.mapExpressions { fst =>
        log("first level: ", fst)
        fst.mapExpressions { snd =>
          log("  second level: ", snd)
          snd
        }
      }
    }
  }
```

At the end of `TailCallTest.scala`. The IR for prefix looks like this:
![prefix](https://github.com/user-attachments/assets/0efa70da-64b5-4910-8875-18dedecfdd94)

As you can see, the prefix direct children is `Literal` and `CallArgument.Specified`. But `CallArgument.Specified` is skipped in mapExpressions:
```
first level: Literal:x
first level: Prefix:((((y) 2)) ((z) 1))
  second level: Prefix:((y) 2)
  second level: Prefix:((z) 1)
```

The relevant code is `arguments.map(_.mapExpressions(fn))` in [Application.Prefix.mapExpressions](https://github.com/enso-org/enso/blob/5d19be472b1a23aa97d7b5cefabe4483163fc500/engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/expression/Application.scala#L144)

**TL;DR;** We cannot rely on `mapExpressions` in `withNewChildren`. It may not be recursive, but it *skips* some IR elements and maps the function on their children (Prefix skips CallArgument.Specified and calls fn on their values).

**Comment by JaroslavTulach** [comment]
OK, that's indeed a different problem than _being recursive_:

> We cannot rely on `mapExpressions` in `withNewChildren`. ... it _skips_ some IR elements and maps the function on their children (Prefix skips CallArgument.Specified and calls fn on their values).

Yes, obviously `mapExpression` deals only with `IR.Expression` and not with the rest of children. But maybe we want to traverse just `Expression` as [stated here](https://github.com/enso-org/enso/pull/11191/files#r1787612365).

**Comment by Akirathan** [comment]
Closed by #11191. Follow-up is #11326.
